hail/python/hailtop/batch/functions.py

```diff
@@ -0,0 +1,111 @@
+import math
+
+from typing import List
+
+from ..utils.utils import grouped
+from .utils import BatchException
+from .batch import Batch
+from .resource import ResourceGroup, ResourceFile
+
+
+def concatenate(b: Batch, files: List[ResourceFile], branching_factor: int = 100):
+    """
+    Concatenate files using tree aggregation.
+
+    Notes
+    -----
+    Runs the `cat` command aggregating files using the `branching_factor`.
+
+    Examples
+    --------
+    Create and execute a batch that concatenates output files:
+
+    >>> b = Batch()
+    >>> j1 = b.new_job()
+    >>> j1.command(f'touch {j1.ofile}')
+    >>> j2 = b.new_job()
+    >>> j2.command(f'touch {j2.ofile}')
+    >>> j3 = b.new_job()
+    >>> j3.command(f'touch {j3.ofile}')
+    >>> files = [j1.ofile, j2.ofile, j3.ofile]
+    >>> ofile = concatenate(b, files, branching_factor=2)
```

Comment:
What do you think of taking the batch from one of the jobs? Maybe also asserting all the jobs have the same batch? That seems like a nicer interface. That would also allow us to make the interface:
```python3
ofile = concatenate(j1.ofile, j2.ofile, j3.ofile, brancing_factor=2)
# or
ofile = concatenate(*files, branching_factor=2)
```

Reply:
I can't really do this right now. I need to add an attribute to the InputResourceFile class to know what batch it came from. But I can't get the cyclical imports to work for the typing of this. I really think in some cases the typing is helpful, but it's an absolute nightmare trying to add new things.